BaragonData/src/main/java/com/hubspot/baragon/utils/UpstreamResolver.java

```diff
@@ -0,0 +1,59 @@
+package com.hubspot.baragon.utils;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Optional;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.net.HostAndPort;
+import com.google.common.net.InetAddresses;
+import com.google.inject.Singleton;
+
+@Singleton
+public class UpstreamResolver {
+  private final Cache<String, String> resolveCache;
+
+  public UpstreamResolver(long maxSize, long expireAfterDays) {
+    this.resolveCache = CacheBuilder.newBuilder()
+        .maximumSize(maxSize)
+        .expireAfterAccess(expireAfterDays, TimeUnit.DAYS)
+        .build();
+  }
+
+  public Optional<String> resolveUpstreamDNS(String address) {
+    try {
+      if (address.contains(":")) {
+        HostAndPort hostAndPort = HostAndPort.fromString(address);
+
+        try {
+          InetAddresses.forString(hostAndPort.getHost());
+          return Optional.of(address); // `address` is already an IP
+        } catch (IllegalArgumentException e) {
+          // `address` is not an IP, continue and try to resolve it
+        }
+
+        String cached = resolveCache.getIfPresent(hostAndPort.getHost());
+        String ip;
+        int port = hostAndPort.getPort();
+        if (cached != null) {
+          ip = cached;
+        } else {
+          ip = InetAddress.getByName(hostAndPort.getHost()).getHostAddress();
+          resolveCache.put(hostAndPort.getHost(), ip);
+        }
+        return Optional.of(String.format("%s:%d", ip, port));
+      } else {
+        try {
+          InetAddresses.forString(address);
```

Comment:
is there a reason we don't try using the cache if it doesn't contain a `:`?

Reply:
I carried this code over from `ResolveUpstreamHelper`, but now you mention it, I don't see a reason for that. Gonna fix that 👍